[FEATURE REQUEST] Auto-refresh when a file is uploaded

JuancaG05 opened the pull request:
## Related Issues
App: https://github.com/owncloud/android/issues/4103

- [x] Added changelog files for the fixed issues in folder changelog/unreleased. More info [here](https://github.com/owncloud/android/tree/master/changelog#create-changelog-items)
_____

## QA

Checklist:

https://github.com/owncloud/android/pull/4199#issuecomment-1784616876

Reports:

- [x] (1) `NEW CONTENT` staying in screen when browsing https://github.com/owncloud/android/pull/4199#issuecomment-1794294182
- [x] (2) `NEW CONTENT` in wrong folder https://github.com/owncloud/android/pull/4199#issuecomment-1798956977

manuelplazaspalacio commented on owncloudApp/src/main/java/com/owncloud/android/presentation/files/filelist/MainFileListFragment.kt:
```diff
@@ -593,6 +598,24 @@ class MainFileListFragment : Fragment(),
                 }
             }
         }
+
```
This should have it's own method.

manuelplazaspalacio commented on owncloudApp/src/main/java/com/owncloud/android/presentation/files/filelist/MainFileListFragment.kt:
```diff
@@ -127,6 +130,7 @@ class MainFileListFragment : Fragment(),
         )
     }
     private val fileOperationsViewModel by sharedViewModel<FileOperationsViewModel>()
+    private val transfersViewModel by viewModel<TransfersViewModel>()
```
This could not be managed with the MainFileListViewModel or by fileOperationsViewModel? I think it's a better solution to not add not view logic to the fragment.

JuancaG05 replied:
Technically it could, but we already have this logic in the TransfersViewModel and:

1) We would need to duplicate code (not a good practice)
2) Semantically, this is something related with transfers so we shouldn't mix it up with other ViewModels

So, taking into account we already have several ViewModels in this fragment, adding another one to prevent duplication of code, I think it's the best option

manuelplazaspalacio reviewed (changes requested):
Some change here @JuancaG05

JuancaG05 commented on owncloudApp/src/main/java/com/owncloud/android/presentation/files/filelist/MainFileListFragment.kt:
```diff
@@ -593,6 +598,24 @@ class MainFileListFragment : Fragment(),
+
+        /* TransfersViewModel observables */
+        // Observe the transfers
+        collectLatestLifecycleFlow(transfersViewModel.transfersWithSpaceStateFlow) { transfers ->
```
If you take a look at the container method, it is `subscribeToViewModels()`, and it is where we add the observers of the different ViewModels in fragments usually. Moving each of the observers out of this method as new methods could be an option, but since they are used just once and `subscribeToViewModels()` has a simple structure, I think we can keep it like that

manuelplazaspalacio replied:
A method should do one thing, this method do a lot of thing. Actually, you had to add some comments to clarify what is done in that part of the code. Don't you think it would be better to isolate that action in a method? You would save on comments, and the code would be more readable, as you wouldn't have to read such a long method.

manuelplazaspalacio reviewed (approved):
LGTM!!

jesmrec commented:
Manual uploads

- [X] 1 file
- [X] 10 files
- [x] 100 files
- [x] 500 files: performance reduced
- [ ] 1000 files
- [ ] 1000 files (outside the target folder)

Other (regression):

- [ ] Auto uploads
- [ ] Uploads from camera
- [ ] Uploads from 3rd party

manuelplazaspalacio reviewed (approved):
LGTM!!

jesmrec commented:
### (1)

1. Upload a number of files >  `max_uploads_to_refresh`
2. When `NEW CONTENT` is displayed, browse to another folder without refreshing. 

Current:  `NEW CONTENT` stays in the screen
Expected: not totally sure... i'd think for the `NEW CONTENT` to update the content of the current folder in case new items where uploaded to that location. If i move to another location, it loses the meaning because this place is updated since it was just browsed. So, i'd go to remove the `NEW CONTENT` if browsing to another place

Pixel 2 Android 11
`e2404875`

JuancaG05 commented:
@jesmrec (1) should be fixed now 👍

jesmrec commented:
i'd raise the default number of uploads to refresh to 200. In terms of performance is affordable and reduce the number of interactions.

jesmrec commented:
### (2)

1. Upload a number of files > max_uploads_to_refresh
2. Just after submitting the upload, browse to another folder, no matter which one and stay there

Current: After some time (when the number of finished uploads overtakes the limit), the fab to refresh is displayed
Expected: Never shown in that folder since this is not the folder to refresh

Pixel 2 Android 11
`d7c56dd61`

JuancaG05 commented:
@jesmrec (2) should be fixed now.
I also updated the maximum number of uploads to refresh automatically to 200, and I introduced a change that I hope helps with performance: manual refresh button will trigger a new refresh only when there is not another one in progress, otherwise it will show a snackbar informing the user there is already a synchronization in progress

jesmrec commented:
This is approved on my side. 

Just as clarification: for a very big amount of files, performance could be affected since device affords too many operations at the time: uploads management including statuses changes, UI updates and long PROPFINFDs parsing that triggers also DB updates. For a reasonable amount of files (1000 - 1500) at the time, this is affordable for a budget device. But, i can not assure that a much bigger amount is going to be performant, depending on the device.

Open to iterate here somehow.
